tests/by-util/test_dirname.rs

```diff
@@ -102,3 +102,135 @@ fn test_emoji_handling() {
+    new_ucmd!()
+        .arg("/foo/bar/baz/.")
+        .succeeds()
+        .stdout_is("/foo/bar/baz\n");
+
+    // Multiple components
+    new_ucmd!()
+        .arg("/usr/local/bin/.")
+        .succeeds()
+        .stdout_is("/usr/local/bin\n");
+}
+
+#[test]
+fn test_trailing_dot_with_zero_flag() {
+    // Test that -z flag works correctly with /. paths
+    new_ucmd!()
+        .arg("-z")
+        .arg("/home/dos/.")
+        .succeeds()
+        .stdout_is("/home/dos\u{0}");
+
+    new_ucmd!()
+        .arg("--zero")
+        .arg("/.")
+        .succeeds()
+        .stdout_is("/\u{0}");
+}
+
+#[test]
+fn test_trailing_dot_multiple_paths() {
+    // Test multiple paths, some with /. suffix
+    new_ucmd!()
+        .args(&["/home/dos/.", "/var/log", "/tmp/."])
+        .succeeds()
+        .stdout_is("/home/dos\n/var\n/tmp\n");
+}
+
+#[test]
+fn test_trailing_dot_edge_cases() {
+    // Double slash before dot (should still work)
+    new_ucmd!()
+        .arg("/home/dos//.")
+        .succeeds()
+        .stdout_is("/home/dos/\n");
+
+    // Path with . in middle (should use normal logic)
+    new_ucmd!()
+        .arg("/path/./to/file")
+        .succeeds()
+        .stdout_is("/path/./to\n");
+}
+
+#[test]
+fn test_trailing_dot_emoji() {
+    // Emoji paths with /. suffix
+    new_ucmd!()
+        .arg("/🌍/path/.")
+        .succeeds()
+        .stdout_is("/🌍/path\n");
+
+    new_ucmd!().arg("/🎉/🚀/.").succeeds().stdout_is("/🎉/🚀\n");
+}
+
+#[test]
+#[cfg(unix)]
+fn test_trailing_dot_non_utf8() {
+    use std::ffi::OsStr;
+    use std::os::unix::ffi::OsStrExt;
+
+    // Create a path with non-UTF-8 bytes ending in /.
+    let non_utf8_bytes = b"/test_\xFF\xFE/.";
+    let non_utf8_path = OsStr::from_bytes(non_utf8_bytes);
+
+    // Test that dirname handles non-UTF-8 paths with /. suffix
+    let result = new_ucmd!().arg(non_utf8_path).succeeds();
+
+    // The output should be the path without the /. suffix
+    let output = result.stdout_str_lossy();
+    assert!(!output.is_empty());
+    assert!(output.contains("test_"));
+    // Should not contain the . at the end
+    assert!(!output.trim().ends_with('.'));
+}
+
+#[test]
+fn test_existing_behavior_preserved() {
+    // Ensure we didn't break existing test cases
+    // These tests verify backward compatibility
+
+    // Normal paths without /. should work as before
+    new_ucmd!().arg("/home/dos").succeeds().stdout_is("/home\n");
+
+    new_ucmd!()
+        .arg("/home/dos/")
+        .succeeds()
+        .stdout_is("/home\n");
+
+    // Parent directory references
+    new_ucmd!()
+        .arg("/home/dos/..")
+        .succeeds()
+        .stdout_is("/home/dos\n");
+
+    // Root
+    new_ucmd!().arg("/").succeeds().stdout_is("/\n");
+
+    // Current directory
+    new_ucmd!().arg(".").succeeds().stdout_is(".\n");
+
+    // Parent reference
+    new_ucmd!().arg("..").succeeds().stdout_is(".\n");
+
+    // Empty string
+    new_ucmd!().arg("").succeeds().stdout_is(".\n");
+}
```

Comment:
I don't understand what the purpose of this test function is. Most of its tests are duplicates. For example, the test for the empty string exists in the `test_empty` function.